chemprop/uncertainty/calibrator.py

```diff
@@ -79,57 +80,220 @@ def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        return
+    r"""Conformalize quantiles to make the interval :math:`[\hat{t}_{\alpha/2}(x),\hat{t}_{1-\alpha/2}(x)]` to have
+    approximately 1 - :math:`\alpha` coverage. [1]_
 
-    def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        ...
-        return
+    .. math::
+        s(x, y) &= \max \left\{ \hat{t}_{\alpha/2}(x) - y, y - \hat{t}_{1-\alpha/2}(x) \right\}
+
+        \hat{q} &= \text{Quantile}(s_1, \ldots, s_n; \left\lceil \frac{(n+1)(1-\alpha)}{n} \right\rceil)
+
+        C(x) &= \left[ \hat{t}_{\alpha/2}(x) - \hat{q}, \hat{t}_{1-\alpha/2}(x) + \hat{q} \right]
 
+    where :math:`s` is the nonconformity score as the difference between y and its nearest quantile.
+    :math:`\hat{t}_{\alpha/2}(x)` and :math:`\hat{t}_{1-\alpha/2}(x)` are the predicted quantiles from quantile
+    regression model.
+
+    .. note::
+        The algorithm is specifically designed for quantile regression model. Intuitively, the set C(x) just
+        grows or shrinks the distance between the quantiles by :math:`\hat{q}` to achieve coverage. However, this
+        function can also be applied to regrssion model without quantiles be provided. In this case, both
+        :math:`\hat{t}_{\alpha/2}(x)` and :math:`\hat{t}_{1-\alpha/2}(x)` are the same as :math:`\hat{y}`. Then, the
+        interval would be the same for every data point (i.e., :math:`\left[-\hat{q}, \hat{q} \right]`).
+
+    Parameters
+    ----------
+    alpha: float
+        The error rate, :math:`\alpha \in [0, 1]`
+
+    References
+    ----------
+    .. [1] Angelopoulos, A.N.; Bates, S.; "A Gentle Introduction to Conformal Prediction and Distribution-Free
+        Uncertainty Quantification." arXiv Preprint 2021, https://arxiv.org/abs/2107.07511
+    """
+
+    def __init__(self, alpha: float):
+        super().__init__()
+        self.alpha = torch.tensor(alpha)
+        self.qhats = torch.tensor([])
+        self.bounds = torch.tensor([-1 / 2, 1 / 2]).view(-1, 1)
 
-@UncertaintyCalibratorRegistry.register("conformal-quantile-regression")
-class ConformalQuantileRegressionCalibrator(UncertaintyCalibrator):
     def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> None:
-        ...
-        return
+        for j in range(preds.shape[1]):
+            mask_j = mask[:, j]
+            targets_j = targets[:, j][mask_j]
+            preds_j = preds[:, j][mask_j]
+            interval_j = uncs[:, j][mask_j]
+
+            interval_bounds = self.bounds * interval_j.unsqueeze(0)
+            pred_bounds = preds_j.unsqueeze(0) + interval_bounds
+
+            calibration_scores = torch.max(pred_bounds[0] - targets_j, targets_j - pred_bounds[1])
+
+            num_data = targets_j.shape[0]
+            q_level = (torch.ceil((num_data + 1) * (1 - self.alpha)) / num_data).item()
+            qhat = torch.quantile(calibration_scores, q_level, interpolation="higher")
+
+            self.qhats = torch.cat((self.qhats, torch.tensor([qhat])))
```

Comment:
That is a good point. I think we generally want to follow the scikit-learn style for this module. To avoid potential errors, we can add a reset function as scikit-learn does ([here](https://github.com/scikit-learn/scikit-learn/blob/70fdc843a4b8182d97a3508c1a426acc5e87e980/sklearn/preprocessing/_data.py#L416)). The value of making it an instance method is that it can return `self` so user can use two different ways to get the parameterized calibrators, whereas using @classmethod can only use the second way. 
1. `calibrator.fit()`
2. `calibrator = .fit()`

TBH, I don't think it matters much as long as we document it well. If you think make it a `@classmethod` would be a better option than the other one, I would be happy to make this change in #986.

Reply:
I made two changes here:
1. use a list instead of tensor for `qhats` and convert it to a tensor after loop
2. generate `qhats` when `.fit()` is called to avoid error